katdal/test/test_applycal.py

```diff
@@ -0,0 +1,164 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###############################################################################
+
+"""Tests for :py:mod:`katdal.applycal`."""
+from __future__ import print_function, division, absolute_import
+from builtins import object, range
+
+import numpy as np
+from numpy.testing import assert_array_equal
+from nose.tools import assert_raises
+import dask.array as da
+
+from katdal.sensordata import SensorCache
+from katdal.categorical import CategoricalData
+from katdal.lazy_indexer import DaskLazyIndexer
+from katdal.applycal import (calc_delay_correction, add_applycal_sensors,
+                             calc_correction_per_corrprod,
+                             apply_vis_correction, add_applycal_transform)
+
+
+POLS = ['v', 'h']
+ANTS = ['m000', 'm001', 'm002', 'm003']
+CENTRE_FREQ = 1284.0
+BANDWIDTH = 856.0
+N_CHANS = 128
+N_DUMPS = 100
+SAMPLE_RATE = 1712.
+
+INPUTS = [ant + pol for ant in ANTS for pol in POLS]
+FREQS = CENTRE_FREQ + BANDWIDTH / N_CHANS * (np.arange(N_CHANS) - N_CHANS // 2)
+INDEX1, INDEX2 = np.triu_indices(len(INPUTS))
+CORRPRODS = [(INPUTS[i1], INPUTS[i2]) for i1, i2 in zip(INDEX1, INDEX2)]
+N_CORRPRODS = len(CORRPRODS)
+
+
+def create_sensor_cache():
+    """Create a SensorCache for testing applycal sensors."""
+    cache = {}
+    delays = np.arange(len(ANTS))
+    delays = np.array([delays, -delays]) / SAMPLE_RATE
+    cache['cal_product_K'] = CategoricalData([np.zeros_like(delays), delays],
+                                             events=[0, 10, N_DUMPS])
+    return SensorCache(cache, timestamps=np.arange(N_DUMPS, dtype=float),
+                       dump_period=1.)
+
+
+def delay_gains(pol, ant):
+    delay = (-1) ** pol * ant
+    return np.exp(2j * np.pi * delay / SAMPLE_RATE * FREQS).astype('complex64')
+
+
+def gains_per_corrprod(dumps, channels):
+    input_map = {ant + pol: (pol_idx, ant_idx)
+                 for (pol_idx, pol) in enumerate(POLS)
+                 for (ant_idx, ant) in enumerate(ANTS)}
+    gains_per_input = np.array([delay_gains(*input_map[inp])
+                                for inp in INPUTS])
+    gains_per_input = gains_per_input[:, channels]
+    gain1 = gains_per_input[INDEX1].T
+    gain2 = gains_per_input[INDEX2].T
+    return gain1 * gain2.conj()
+
+
+class TestCorrectionPerInput(object):
+    """Test the :func:`~katdal.applycal.calc_*_correction` functions."""
+    def setup(self):
+        self.cache = create_sensor_cache()
+
+    def test_calc_delay_correction(self):
+        for n in range(len(ANTS)):
+            for m in range(len(POLS)):
+                sensor = calc_delay_correction(self.cache, 'K', (m, n), FREQS)
+                assert_array_equal(sensor[n],
+                                   np.ones(N_CHANS, dtype='complex64'))
+                assert_array_equal(sensor[10 + n], delay_gains(m, n))
```

Comment:
Yes, just to introduce some extra variation in the test. Since K is currently not interpolated, all dumps from 10 onwards should deliver the same delay correction sensor value.